Bugfix: Add proper padding when scrolling is added to bar gauge

mckn opened the pull request:
**What this PR does / why we need it**:
Prior to this PR the scrollbar will hide the values in the bar gauge (and probably other visualizations). I added a setScrollable to apply the scrolling in the PanelChrome instead of in the `vizRepeater` which results in getting the padding in the right place for all browsers.

**Which issue(s) this PR fixes**:
Fixes #32083

**Special notes for your reviewer**:
I'm a bit concerned about adding the `overflow: hidden` to the `.panel-content`. Have done some testing and think it should  be fine but still changing that component feels a bit dangerous.

mckn committed:
fixing so we get the scrollbar in the proper place for the vizualization.

mckn commented on packages/grafana-ui/src/components/VizRepeater/VizRepeater.tsx:
```diff
@@ -152,7 +152,6 @@ export class VizRepeater<V, D = {}> extends PureComponent<Props<V, D>, State<V>>
 
     const repeaterStyle: React.CSSProperties = {
       display: 'flex',
-      overflow: minVizHeight ? 'hidden auto' : 'hidden',
```
Note: this should be fine since I have added the `overflow: hidden` to the `.panel-content` and the only consumer of the `VizRepater` that use the `minVizHeight` was the bar gauge.

mckn commented on public/sass/pages/_dashboard.scss:
```diff
@@ -102,11 +102,22 @@ div.flot-text {
   width: 100%;
   flex-grow: 1;
   height: calc(100% - #{$panel-header-height});
+  overflow: hidden;
```
Note: this feels a bit dangerous. Do you know about any use cases that might introduce a bug by adding this code? @torkelo

torkelo replied:
it breaks the dropdown select in Table panel, if we had a proper portal for Select component that could solve that

mckn replied:
hmm, damn...so I need to rethink this a bit.

mckn replied:
Let me know if you think this is something we should proceed with. I will wait with fixing the tests until I got a 👍🏻

mckn replied:
And I guess we want to support horizontal scrolling as well.

mckn commented on public/sass/pages/_dashboard.scss:
```diff
@@ -102,11 +102,22 @@ div.flot-text {
+    margin-bottom: $panel-padding;
+  }
+
+  &--no-padding.panel-content--scrollable-y {
```
Note: can I write this in some better way using sass?

Elfo404 replied:
```suggestion
  &--no-padding#{&}--scrollable-y {
```

might do the trick :)

mckn committed:
changed so the select is used with a portal.

mckn committed:
increasing the zIndex of the selector.

mckn commented:
***Summary:***
When I changed the `.panel-content` to have `overflow: hidden` the selector stopped working e.g. in the scenario below. I fixed that by changing so we render the select menu in a portal.

<img width="917" alt="Skärmavbild 2021-03-31 kl  10 49 39" src="https://user-images.githubusercontent.com/172951/113117602-e7bc3200-920e-11eb-8f2d-1274037c1626.png">

However that resulted in the following issue in scenarios where we have a selector in a modal. That was solved by increasing the zIndex of the select menu portal to same as the modal.

<img width="934" alt="Skärmavbild 2021-03-31 kl  10 51 18" src="https://user-images.githubusercontent.com/172951/113117733-091d1e00-920f-11eb-9e5b-caed59e26fda.png">

The benefit of doing this is that we also squash a bunch of bugs where we use the selector in e.g. transformations etc where we currently have forced the selector to open downwards to prevent it from being hidden. But again, increasing the zIndex feels like a slippery slope.

I can think of one scenario (not sure if it is a valid one) where having the same (or higher) zIndex of the select menu portal and the modal is if we want to have an open select but still present a modal on top of it.

Would love some feedback on this @grafana/grafana-frontend-platform 🙏🏻

torkelo commented on public/app/features/dashboard/dashgrid/PanelChrome.tsx:
```diff
@@ -274,10 +274,12 @@ export class PanelChrome extends Component<Props, State> {
     const chromePadding = plugin.noPadding ? 0 : theme.panelPadding;
     const panelWidth = width - chromePadding * 2 - PANEL_BORDER;
-    const innerPanelHeight = height - headerHeight - chromePadding * 2 - PANEL_BORDER;
+    const scrollableYMargin = plugin.scrollableY ? theme.panelPadding : 0;
+    const innerPanelHeight = height - headerHeight - chromePadding * 2 - PANEL_BORDER - scrollableYMargin;
     const panelContentClassNames = classNames({
       'panel-content': true,
       'panel-content--no-padding': plugin.noPadding,
+      'panel-content--scrollable-y': plugin.scrollableY,
```
Not sure we need this special logic. and If plugin.scrollable is true we should wrap the component in a CustomScrollbar

torkelo committed:
Merge remote-tracking branch 'origin/master' into mckn/bar-gauge-scroll

torkelo committed:
Testing custom scrollbar

torkelo committed:
Merge branch 'mckn/bar-gauge-scroll' of github.com:grafana/grafana into mckn/bar-gauge-scroll

torkelo committed:
Progress but now DashList has no right padding

torkelo committed:
added back className

mckn committed:
wip

mckn committed:
fixing the issues with zIndexes and having the select as a portal.

mckn committed:
adding portal to theme definition.

jackw committed:
test(selectbase): migrate tests to RTL and fix for portal related changes

dprokop commented:
@mckn this fixes the bar gauge issue 👌  But I see that adding portaling to select introduces the exact issues me and @tskarhed has been struggling with when we tried to add portaling:

1. There's sth wrong with Table's panel select:
  ![Kapture 2021-04-08 at 09 17 17](https://user-images.githubusercontent.com/2376619/113984980-f4163f80-984b-11eb-91f1-3b4b420dd213.gif)


2. The panel edit selects are now positioned incorrectly, add scrollbar to the viewport and have fixed position relative to the window when scrolling. This applies to query editor selects and other selects as well
  ![Kapture 2021-04-08 at 09 18 25](https://user-images.githubusercontent.com/2376619/113984946-ed87c800-984b-11eb-9e12-448f62e8f020.gif)
  ![Kapture 2021-04-08 at 09 24 10](https://user-images.githubusercontent.com/2376619/113985274-448d9d00-984c-11eb-8f12-64a97ede5790.gif)


3. The initial position of the select is incorrect:
  ![Kapture 2021-04-08 at 09 19 00](https://user-images.githubusercontent.com/2376619/113984888-df39ac00-984b-11eb-8486-056c9906d831.gif)

tskarhed commented:
Wohoo! Select issues, don't you just love them? 😁

mckn commented:
> @mckn this fixes the bar gauge issue 👌 But I see that adding portaling to select introduces the exact issues me and @tskarhed has been struggling with when we tried to add portaling:
> 
> 1. There's sth wrong with Table's panel select:
>    ![Kapture 2021-04-08 at 09 17 17](https://user-images.githubusercontent.com/2376619/113984980-f4163f80-984b-11eb-91f1-3b4b420dd213.gif)
> 2. The panel edit selects are now positioned incorrectly, add scrollbar to the viewport and have fixed position relative to the window when scrolling. This applies to query editor selects and other selects as well
>    ![Kapture 2021-04-08 at 09 18 25](https://user-images.githubusercontent.com/2376619/113984946-ed87c800-984b-11eb-9e12-448f62e8f020.gif)
>    ![Kapture 2021-04-08 at 09 24 10](https://user-images.githubusercontent.com/2376619/113985274-448d9d00-984c-11eb-8f12-64a97ede5790.gif)
> 3. The initial position of the select is incorrect:
>    ![Kapture 2021-04-08 at 09 19 00](https://user-images.githubusercontent.com/2376619/113984888-df39ac00-984b-11eb-8486-056c9906d831.gif)

Awesome findings @dprokop! I think for some of this issues we probably need to expose a "usePortal" property on the select component so you can control that depending on the context. It doesn't feel good but it might be the way to go.

mckn commented:
Closing this one in favor of #32835 and #32833
